src/renderer/dx/CustomTextLayout.cpp

```diff
@@ -1379,3 +1529,229 @@ void CustomTextLayout::_OrderRuns()
+                // translation moving up (taking excess descent and adding to ascent) and positive is the opposite.
+                const auto boxVerticalTranslationDesignUnits = (extraAscent - extraDescent) / 2;
+
+                // The translation is just a raw movement of pixels up or down. Since we were working in Design Units,
+                // we need to run the opposite algorithm shown above to go from Design Units to Pixels.
+                boxVerticalTranslation = boxVerticalTranslationDesignUnits * scaledFontSize / fontMetrics.designUnitsPerEm;
+            }
+        }
+    }
+
+    // The horizontal adjustments follow the exact same logic as the vertical ones.
+    float boxHorizontalScaleFactor = 1.0f;
+    float boxHorizontalTranslation = 0.0f;
+    {
+        // This is the only difference. We don't have a horizontalOriginX from the metrics.
+        // However, https://docs.microsoft.com/en-us/windows/win32/api/dwrite/ns-dwrite-dwrite_glyph_metrics says
+        // the X coordinate is specified by half the advanceWidth to the right of the horizontalOrigin.
+        // So we'll use that as the "center" and apply it the role that verticalOriginY had above.
+        const auto boxCenter = boxMetrics.advanceWidth / 2.0f;
+
+        const auto boxLeftDesignUnits = boxCenter - boxMetrics.leftSideBearing;
+        const auto boxRightDesignUnits = boxMetrics.advanceWidth - boxMetrics.rightSideBearing - boxCenter;
+        const auto boxWidthDesignUnits = boxLeftDesignUnits + boxRightDesignUnits;
+
+        const auto cellLeftDesignUnits = boxCenter;
+        const auto cellRightDesignUnits = boxCenter;
+        const auto cellWidthDesignUnits = cellLeftDesignUnits + cellRightDesignUnits;
+
+        const auto boxTouchesCellLeft = boxLeftDesignUnits >= cellLeftDesignUnits;
+        const auto boxTouchesCellRight = boxRightDesignUnits >= cellRightDesignUnits;
+        const auto boxIsWideEnoughForCell = boxWidthDesignUnits >= cellWidthDesignUnits;
+
+        if (!(boxTouchesCellLeft && boxTouchesCellRight && boxIsWideEnoughForCell))
+        {
+            {
+                boxHorizontalScaleFactor = cellWidthDesignUnits / boxWidthDesignUnits;
+            }
+            {
+                const auto extraLeft = boxLeftDesignUnits * boxHorizontalScaleFactor - cellLeftDesignUnits;
+                const auto extraRight = boxRightDesignUnits * boxHorizontalScaleFactor - cellRightDesignUnits;
+
+                const auto boxHorizontalTranslationDesignUnits = (extraLeft - extraRight) / 2;
+
+                boxHorizontalTranslation = boxHorizontalTranslationDesignUnits * scaledFontSize / fontMetrics.designUnitsPerEm;
+            }
+        }
+    }
+
+    // OK, make the object that will represent our effect, stuff the metrics into it, and return it.
+    RETURN_IF_FAILED(WRL::MakeAndInitialize<BoxDrawingEffect>(effect, boxVerticalScaleFactor, boxVerticalTranslation, boxHorizontalScaleFactor, boxHorizontalTranslation));

```

Comment:
should we consider treating the _needs no modifications_ state by having _no BoxDrawingEffect_, and letting the standard glyph run drawing handle it? if the font needs no adjustment we could also simply _not split the run_


(we may want to keep splitting the run if we ever want to use the "use line and fill primitives" idea from other terminals, or the exploded-bitmap version from VTE)

Reply:
> by having _no BoxDrawingEffect_

Oh I like this idea a lot

Reply:
I dropped the effect. Not splitting the run might make the code more complicated than we want right now. We can revisit if it turns out to make a perf problem, but splitting runs all day is basically what this analyzer does anyway.